Microsoft.HealthVault.Client.NetFramework/NetFrameworkSecretStore.cs

```diff
@@ -0,0 +1,113 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.HealthVault.Client.Core;
+
+namespace Microsoft.HealthVault.Client
+{
+    internal class NetFrameworkSecretStore : ISecretStore
+    {
+        private const string SaltString = "d219a6c4e0a78d74f2ff309061d473424646e43fc72f88aca5801ff09c33da9c";
+
+        private static readonly byte[] Salt = StringToByteArray(SaltString);
+
+        private static readonly string StoreFolder = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), HealthVaultConstants.Storage.DirectoryName);
+
+        public async Task WriteAsync(string key, string contents)
+        {
+            try
+            {
+                string filePath = Path.Combine(StoreFolder, key);
+
+                EnsureDirectory();
+
+                byte[] contentBytes = Encoding.UTF8.GetBytes(contents);
+                byte[] encryptedContentBytes = ProtectedData.Protect(contentBytes, Salt, DataProtectionScope.CurrentUser);
+
+                using (FileStream fileStream = new FileStream(filePath, FileMode.Create, FileAccess.Write))
+                {
+                    await fileStream.WriteAsync(encryptedContentBytes, 0, encryptedContentBytes.Length).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e) when (!(e is IOException))
+            {
+                throw new IOException(string.Format(ClientResources.FileAccessErrorMessage, ClientResources.FileAccessActionWrite, key), e);
+            }
+        }
+
+        public async Task<string> ReadAsync(string key)
+        {
+            try
+            {
+                string filePath = Path.Combine(StoreFolder, key);
+
+                using (FileStream fileStream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
+                {
+                    byte[] encryptedContentBytes = new byte[fileStream.Length];
+                    await fileStream.ReadAsync(encryptedContentBytes, 0, (int)fileStream.Length).ConfigureAwait(false);
+                    byte[] contentBytes = ProtectedData.Unprotect(encryptedContentBytes, Salt, DataProtectionScope.CurrentUser);
+
+                    return Encoding.UTF8.GetString(contentBytes);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                return null;
+            }
+            catch (DirectoryNotFoundException)
+            {
+                return null;
+            }
```

Comment:
Other platforms implement the secret store with different APIs, only UWP and .NET Framework use files. The caller shouldn't need to know about that implementation and in our case the directory not existing simply means that the requested secret doesn't exist. Which is communicated by returning a null. Do you think we should throw a SecretNotFoundException in this case, with the DirectoryNotFoundException as an inner exception?

---
In reply to: [121512775](https://github.com/Microsoft/healthvault-dotnetstandard-sdk/pull/52#discussion_r121512775) [](ancestors = 121512775)